### PR TITLE
Micro-optimize `InferCtxt::replace_bound_vars_with_placeholders` by avoiding unnecessary allocation

### PR DESCRIPTION
**Pull request by martingms**
Showed up hot in a callgrind profile of compiling `bitmaps-3.1.0`

`replace_bound_vars` allocates a `BTreeMap` that was only used for an `.is_empty()` check. This PR replaces the map with a simple `boolean`.

Local benchmarks:

#### Primary benchmarks

Benchmark | Profile | Scenario | % Change | Significance Factor?
-- | -- | -- | -- | --
diesel-1.4.8 | check | full | -0.38% | 1.91x
diesel-1.4.8 | debug | full | -0.37% | 1.87x
diesel-1.4.8 | opt | full | -0.37% | 1.85x
diesel-1.4.8 | check | incr-full | -0.34% | 1.72x
diesel-1.4.8 | debug | incr-full | -0.33% | 1.67x
diesel-1.4.8 | opt | incr-full | -0.32% | 1.61x

#### Secondary benchmarks

Benchmark | Profile | Scenario | % Change | Significance Factor?
-- | -- | -- | -- | --
coercions | debug | full | 1.96% | 9.81x
bitmaps-3.1.0 | check | full | -1.54% | 7.72x
bitmaps-3.1.0 | debug | full | -1.48% | 7.38x
bitmaps-3.1.0 | opt | full | -1.39% | 6.94x
bitmaps-3.1.0 | check | incr-full | -1.26% | 6.30x
bitmaps-3.1.0 | debug | incr-full | -1.19% | 5.97x
bitmaps-3.1.0 | opt | incr-full | -1.15% | 5.73x
tt-muncher | opt | full | -0.50% | 2.50x
wf-projection-stress-65510 | opt | full | -0.49% | 2.47x
wf-projection-stress-65510 | check | full | -0.44% | 2.22x
wf-projection-stress-65510 | debug | full | -0.42% | 2.12x
wf-projection-stress-65510 | check | incr-full | -0.40% | 2.02x
deep-vector | debug | incr-patched: add vec item | -0.40% | 1.98x
deep-vector | opt | incr-patched: println | 0.39% | 1.93x
wf-projection-stress-65510 | debug | incr-full | -0.38% | 1.92x
wf-projection-stress-65510 | opt | incr-full | -0.37% | 1.84x
projection-caching | check | full | -0.37% | 1.83x
deep-vector | debug | full | -0.32% | 1.61x

r? @nnethercote

### CODE REVIEW

File: compiler/rustc_infer/src/infer/higher_ranked/mod.rs [hunk 1]
```diff
@@ -100,11 +103,12 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             })
         };
 
-        let (result, map) = self.tcx.replace_bound_vars(binder, fld_r, fld_t, fld_c);
+        let result =
+            self.tcx.replace_escaping_bound_vars(binder.skip_binder(), fld_r, fld_t, fld_c);
 
         // If there were higher-ranked regions to replace, then actually create
         // the next universe (this avoids needlessly creating universes).
-        if !map.is_empty() {
```
Comment (jackh726) [inline compiler/rustc_infer/src/infer/higher_ranked/mod.rs:107]: Wait...is this wrong?

I think this should be taking into account types and consts too...

Comment (compiler-errors) [reply compiler/rustc_infer/src/infer/higher_ranked/mod.rs:107]: Do we see bound types in practice? Like other than in canonical queries, which I assume we liberate elsewhere? (maybe not)

Comment (jackh726) [reply compiler/rustc_infer/src/infer/higher_ranked/mod.rs:107]: We don't :) At least not much today. Which is likely why this isn't doesn't fail everywhere. But in the future, we'll probably see more.

Comment (nnethercote) [reply compiler/rustc_infer/src/infer/higher_ranked/mod.rs:107]: Is the wrongness pre-existing? This change looks ok to me from a mechanical refactoring point of view, but I don't know much about this code in general.

Comment (jackh726) [reply compiler/rustc_infer/src/infer/higher_ranked/mod.rs:107]: Yeah, looks preexisting.

Comment (nnethercote) [reply compiler/rustc_infer/src/infer/higher_ranked/mod.rs:107]: So do you have any concern with this PR being merged?

Comment (jackh726) [reply compiler/rustc_infer/src/infer/higher_ranked/mod.rs:107]: Nope! But someone (probably me) should file an issue so this doesn't get forgotten.

**Comment by nnethercote** [comment]
Another weird performance result... I don't know what's happening. This change really seems like it should be a straight win.

**Comment by nnethercote** [comment]
#97345 landed recently. It entirely removes a great chunk of the compiler's work for `bitmaps` and similar benchmarks, including the calls to `InferCtxt::replace_bound_vars_with_placeholders`. So this PR isn't needed any more. Thanks again for the attempt, @martingms!

**Pull request closed without merging**